### PR TITLE
feat(sdk): OpenMetrics support for System Metrics

### LINKED ISSUE #4903: [CLI]: SystemMonitor thread starting new thread, None as the result
### Describe the bug

Kernel run by jupyter/jupyter-lab. Login succesfull. Run was logged on wandb.ai - I hope that without distortions.
Happens on first run and any next. 

```
Exception in thread SystemMonitor:
Traceback (most recent call last):
  File "/home/mkurtys/.pyenv/versions/3.10.9/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/home/mkurtys/.pyenv/versions/3.10.9/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/home/mkurtys/.pyenv/versions/3.10.9/envs/torch/lib/python3.10/site-packages/wandb/sdk/internal/system/system_monitor.py", line 118, in _start
    asset.start()
  File "/home/mkurtys/.pyenv/versions/3.10.9/envs/torch/lib/python3.10/site-packages/wandb/sdk/internal/system/assets/cpu.py", line 166, in start
    self.metrics_monitor.start()
  File "/home/mkurtys/.pyenv/versions/3.10.9/envs/torch/lib/python3.10/site-packages/wandb/sdk/internal/system/assets/interfaces.py", line 168, in start
    logger.info(f"Started {self._process.name}")
AttributeError: 'NoneType' object has no attribute 'name'
Exception in thread SystemMonitor:
Traceback (most recent call last):
  File "/home/mkurtys/.pyenv/versions/3.10.9/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/home/mkurtys/.pyenv/versions/3.10.9/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/home/mkurtys/.pyenv/versions/3.10.9/envs/torch/lib/python3.10/site-packages/wandb/sdk/internal/system/system_monitor.py", line 118, in _start
    asset.start()
  File "/home/mkurtys/.pyenv/versions/3.10.9/envs/torch/lib/python3.10/site-packages/wandb/sdk/internal/system/assets/cpu.py", line 166, in start
    self.metrics_monitor.start()
  File "/home/mkurtys/.pyenv/versions/3.10.9/envs/torch/lib/python3.10/site-packages/wandb/sdk/internal/system/assets/interfaces.py", line 168, in start
    logger.info(f"Started {self._process.name}")
AttributeError: 'NoneType' object has no attribute 'name'
```

### Additional Files

_No response_

### Environment

WandB version: wandb==0.13.9

OS: x86_64 GNU/Linux

Python version: Python 3.10.9  [GCC 12.2.0] on linux

Versions of relevant libraries:

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dmitryduev**
Fixes WB-NNNN
Fixes #4903

Description
-----------
This PR implements initial support of OpenMetrics endpoint consumption for our System Metrics.
This should work in the general case, but is specifically targeted at the NVIDIA DCGM Exporter.

image

image

Notes:
- For now we will only support the Counter and Gauge Prometheus metric types.
- Added telemetry to track usage.
- New setting for OpenMetrics endpoints to be monitored (Map string -> string -- tag -> uri)
- The metrics are posted in the following format: `system/openmetrics.<tag>.<metric name>.<label index>` (to ease name-spacing on the FE). OM labels are hashed and associated with an index to adapt it to the way we store and display the system metrics. For example, the following OM data

```
# HELP DCGM_FI_DEV_MEM_COPY_UTIL Memory utilization (in %).
# TYPE DCGM_FI_DEV_MEM_COPY_UTIL gauge
DCGM_FI_DEV_MEM_COPY_UTIL{{gpu="0",UUID="GPU-c601d117-58ff-cd30-ae20-529ab192ba51",device="nvidia0",modelName="Tesla T4",Hostname="gke-gke-dcgm-default-pool-eb7746d2-6vkd",container="",namespace="",pod=""}} 10.4
DCGM_FI_DEV_MEM_COPY_UTIL{{gpu="1",UUID="GPU-a7c8aa83-d112-b585-8456-5fc2f3e6d18e",device="nvidia1",modelName="Tesla T4",Hostname="gke-gke-dcgm-default-pool-eb7746d2-6vkd",container="dcgm-loadtest",namespace="default",pod="dcgm-loadtest"}} 99.1
```

will be converted to `{"system/openmetrics.DCGM.DCGM_FI_DEV_MEM_COPY_UTIL.0": 10.4, "system/openmetrics.DCGM.DCGM_FI_DEV_MEM_COPY_UTIL.1": 99.1}`. We could store the label hash -> labels mapping (todo for a future pr.)
- TODO: in a future PR, need to figure out how to do job (e.g. cluster job scheduled with SLURM or something) name-spacing

Testing
-------
- [x] Unit test
- [x] Fake flask-based local prometheus endpoint
- [x] Multi-node multi-GPU GKE cluster with DCGM/DCGM Exporter deployed there

Checklist
-------
- [x] Include reference to internal ticket "Fixes WB-NNNN" and/or GitHub issue "Fixes #NNNN" (if applicable)
- [x] Ensure PR title compliance with the [conventional commits standards](https://github.com/wandb/wandb/blob/main/CONTRIBUTING.md#conventional-commits)

### CODE REVIEW

File: wandb/sdk/internal/system/assets/open_metrics.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+            wandb.termwarn(
+                "Monitoring OpenMetrics endpoints requires the `prometheus_client` package. "
+                "To install it, run `pip install prometheus_client`.",
+                repeat=False,
+            )
+            return False
+        # check if the endpoint is available and is a valid OpenMetrics endpoint
+        try:
+            assert prometheus_client_parser is not None
+            response = requests.get(url)
```
[the remark this answers is not in the document]
Comment (raubitsj) [reply wandb/sdk/internal/system/assets/open_metrics.py:153]: can you remind me, is_available is maps to what process/thread context.   is this a sys metrics thread, or in the context of handling the run_start message?

File: wandb/sdk/internal/system/assets/open_metrics.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    def teardown(self) -> None:
+        if self.session is not None:
+            self.session.close()
+            self.session = None
+
+    def parse_open_metrics_endpoint(self) -> Dict[str, Union[str, int, float]]:
+        assert prometheus_client_parser is not None
+        assert self.session is not None
+
+        response = self.session.get(self.url)
```
Comment (raubitsj) [inline wandb/sdk/internal/system/assets/open_metrics.py:63]: are you ok with this not having a timeout?

**Review by raubitsj — COMMENTED** [review]
Two overall concerns are:
1. hardening of network paths (timeouts?) especially during probe
2. possibility of pulling too many metrics

for 1:
I think i would be less concerned if this was opt-in for the first release.  With the auto probe it can cause unintended consequences. for example if you have a web service responding to "http://localhost:9400/metrics" but not returning data,  this will hang is_available... not sure about the downstream consequences of that.

for 2:
also less of a concern if this is opt in for first release.

File: wandb/sdk/internal/system/assets/open_metrics.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmitryduev) [reply wandb/sdk/internal/system/assets/open_metrics.py]: re: context -- it's the latter. (we check if stuff is available once when starting the system monitor in handle_request_run_start)

**Comment by dmitryduev** [comment]
@raubitsj I hardened the network paths, as requested, and made this feature an opt-in one.

File: wandb/sdk/internal/system/assets/open_metrics.py (hunk 2 above)
Comment (dmitryduev) [reply wandb/sdk/internal/system/assets/open_metrics.py]: this should not be a possibility anymore

**Review by raubitsj — APPROVED** [review]
LGTM

I think there is still some network things to think about, but since this is opt in, it should be fine.   What happens if a promethius endpoint doesnt return any data and hangs a http request?   I think this will cause the metrics probe to hang, which will block the run from starting.  (i didnt see any timeouts in the session setup, but there might be defaults).   In the rest of our networking code we set a default timeout and do retries.

**Pull request merged**